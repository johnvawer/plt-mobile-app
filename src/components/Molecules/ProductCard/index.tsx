import React from 'react'
import {
  View,
  TouchableOpacity,
  Text,
  Image,
  StyleSheet,
  useWindowDimensions
} from 'react-native'
import Button from '../../Atoms/Button'
import ProductPrice from '../../Atoms/ProductPrice'

interface Props {
  onPress: () => void
  product: Product
  buttonText: string
}

const ProductCard = ({ product, onPress, buttonText }: Props): JSX.Element => {
  const { width } = useWindowDimensions()
  const { name, price, img, id } = product
  const secureImage = img.split('http://').join('https://')

  return (
    <TouchableOpacity onPress={onPress} key={id}>
      <View style={styles.container}>
        <Image
          testID='productCardImage'
          resizeMode='contain'
          source={{ uri: secureImage }}
          style={[styles.image, {
            width: width - 20,
            height: width * 1.25
          }]}
        />
        <Text>{name}</Text>
        <ProductPrice amount={price} />
        <Button onPress={onPress} buttonText={buttonText} />
      </View>
    </TouchableOpacity>
  )
}

const styles = StyleSheet.create({
  container: {
    margin: 10,
    alignItems: 'center'
  },
  image: {
    marginBottom: 10
  }
})

export default ProductCard
