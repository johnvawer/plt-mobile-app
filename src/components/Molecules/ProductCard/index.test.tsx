import React from 'react'
import { render, fireEvent } from '@testing-library/react-native'
import ProductCard from '.'

describe('<ProductCard />', () => {
  let props: {
    onPress: () => void
    product: Product
    buttonText: string
  }
  let onPress: jest.Mock

  beforeEach(() => {
    onPress = jest.fn()
    props = {
      onPress,
      product: {
        id: 1,
        name: 'Black Dress',
        colour: 'Black',
        img: 'http://www.image.com/someimage.png',
        price: 10
      },
      buttonText: 'My Button Text'
    }
  })

  afterEach(() => {
    onPress.mockRestore()
  })

  describe('<ProductPrice />', () => {
    it('should take the props and render them correctly', () => {
      const { toJSON } = render(<ProductCard {...props} />)

      expect(toJSON()).toMatchSnapshot()
    })

    it('should call the onPress prop on button press', () => {
      const { getByText } = render(<ProductCard {...props} />)
      const button = getByText('My Button Text')
      fireEvent.press(button)

      expect(onPress).toHaveBeenCalledTimes(1)
    })

    it('should replace insecure image protocols', () => {
      const { getByTestId } = render(<ProductCard {...props} />)
      const image = getByTestId('productCardImage')

      expect(image.props.source.uri).toEqual('https://www.image.com/someimage.png')
    })
  })
})
