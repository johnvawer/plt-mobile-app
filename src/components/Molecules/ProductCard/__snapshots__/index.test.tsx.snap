// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProductCard /> <ProductPrice /> should take the props and render them correctly 1`] = `
<View
  accessible={true}
  collapsable={false}
  focusable={true}
  nativeID="animatedComponent"
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "opacity": 1,
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "margin": 10,
      }
    }
  >
    <Image
      resizeMode="contain"
      source={
        Object {
          "uri": "https://www.image.com/someimage.png",
        }
      }
      style={
        Array [
          Object {
            "marginBottom": 10,
          },
          Object {
            "height": 937.5,
            "width": 730,
          },
        ]
      }
      testID="productCardImage"
    />
    <Text>
      Black Dress
    </Text>
    <Text
      testID=""
    >
      Â£
      10.00
    </Text>
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "white",
          "borderColor": "black",
          "borderWidth": 1,
          "margin": 5,
          "padding": 10,
        }
      }
    >
      <Text>
        My Button Text
      </Text>
    </View>
  </View>
</View>
`;
